### ISSUE TITLE
XvB restart due to failed request leads to too few HR sent for one hour

### ISSUE THREAD

**Issue by Cyrix126** [Cyrix126/gupaxx#26]
## Bug
If only one request to XvB failed for the public stats, XvB algorithm will restart and forget what it sent, sending too few HR since it will think there is external HR and get dropped from the 1 hour average for almost an hour.

## Steps
1. Run XvB with enough HR to participate in a round
2. Wait, if your connection to the XvB drop once when the request is made (every minutes), XvB algo will restart.
3. The algo will send not enough HR while trying to get to the same round for an hour.

## Screenshots
![bug_xvb](https://github.com/user-attachments/assets/826f5cdd-cf33-4503-981d-8d630a4c7996)

terminal output:
```
XvB Watchdog | Could not send HTTP API request to: https://xmrvsbeast.com/p2pool/stats                                                     
:error sending request for url (https://xmrvsbeast.com/p2pool/stats) 
```

## Additional context
XvB algorithm keeps in memory the last hour average HR it sent to XvB, to compare with HR announced on the API to know if there is any external HR.
When it restarts, it forgets about the HR it sent and believes the HR sent on XvB the last hour is only external, making it sending less.

## Solution
Do not forget the last average hour even if the XvB algo is stopped.

**Comment by MKRajan29** [comment]
Hey, how are you able to select the manual donation level?

**Comment by MKRajan29** [comment]
![image](https://github.com/user-attachments/assets/5cdc2e7b-43e2-48bb-aba9-142deeae3a89)

**Comment by MKRajan29** [comment]
Also for the last few days I have been observing that the Round is either Whale or Mega, rarely any other. Is this deliberate or coincidentally random?

**Comment by MKRajan29** [comment]
> Hey, how are you able to select the manual donation level?

Oh, I see, this is on Linux, why is it not on Windows?

**Comment by MKRajan29** [comment]
Also on the XMRvsBeast is this normal in my History. or is it what you are talking about ?

![image](https://github.com/user-attachments/assets/69e9a30e-f253-4db2-be58-2f2b12c95ef8)

**Comment by Cyrix126** [comment]
> Also for the last few days I have been observing that the Round is either Whale or Mega, rarely any other. Is this deliberate or coincidentally random?

https://monero.town/post/4723333

**Comment by Cyrix126** [comment]
> Also on the XMRvsBeast is this normal in my History. or is it what you are talking about ?

Possibly.

**Comment by Cyrix126** [comment]
> Also in these, why does it say "Sending 0 seconds to XvB", is that how it is supposed to say?

Sorry the output in the XvB console is not very clear. You can check the terminal output to have the full logic of the algo for now.

**Comment by Cyrix126** [comment]
To fix this issue, the failed request will be retried multiples times before quitting.
Remembering the HR sent between start of algorithm will be part of a feature to store logs for history purpose.

**Comment by Cyrix126** [comment]
fixed in https://github.com/Cyrix126/gupaxx/commit/a9efa2b2f4dad8b34f95fec79c82c7efabd3ed0b
Issue will be closed at next release.

**Comment by Cyrix126** [comment]
closed by release [v1.5.2](https://github.com/Cyrix126/gupaxx/releases/tag/v1.5.2)
